File: cli/src/modules/compose/devcontainer.ts
import type { ComposeService } from '@/types.js';
import { SSH_DEFAULTS } from '@/ssh-defaults.js';
import {
  DIND_ENGINE_HOST,
  DIND_ENGINE_PORT,
  DIND_ENGINE_NETWORK,
} from '@/modules/compose/dind-engine.js';

export type DockerSocketMode = 'none' | 'dind';

const SOCKET_CHOICES: { value: DockerSocketMode; label: string }[] = [
  { value: 'none', label: 'none — no Docker access (safest)' },
  {
    value: 'dind',
    label: 'dind — isolated rootless Docker-in-Docker engine (sandbox)',
  },
];

export const devcontainerService: ComposeService = {
  id: 'devcontainer',
  label: `${SSH_DEFAULTS.serviceName} (main)`,
  always: true,
  options: [
    {
      id: 'dockerSocket',
      label: 'Docker access mode',
      type: 'select',
      choices: SOCKET_CHOICES,
      default: 'none',
    },
  ],
  render({ imageName, enabledServiceIds, options }) {
    const depends = enabledServiceIds.filter((s) =>
      ['mongo', 'redis', 'postgres'].includes(s),
    );
    const requested = (options.dockerSocket as DockerSocketMode | undefined) ?? 'none';
    const dindEnabled = enabledServiceIds.includes(DIND_ENGINE_HOST);

    // Fail safe: if dind is requested but the engine service is not enabled,
    // fall back to 'none' (no access) rather than producing a broken DOCKER_HOST.
    const mode: DockerSocketMode = requested === 'dind' && dindEnabled ? 'dind' : 'none';

    const volumes: string[] = [
      '../..:/workspace',
      'devcontainer_etc:/etc',
      'devcontainer_root:/root',
      'devcontainer_home:/home',
    ];

    const networks: string[] = ['local-network'];
    if (mode === 'dind') networks.push(DIND_ENGINE_NETWORK);

    const svc: Record<string, unknown> = {
      image: imageName,
      build: '.',
      container_name: SSH_DEFAULTS.serviceName,
      command: 'sleep infinity',
      restart: 'unless-stopped',
      volumes,
      networks,
    };

    if (mode === 'dind') {
      svc.environment = [`DOCKER_HOST=tcp://${DIND_ENGINE_HOST}:${DIND_ENGINE_PORT}`];
    }

    const extraDepends = mode === 'dind' ? [DIND_ENGINE_HOST] : [];
    const allDepends = [...depends, ...extraDepends];
    if (allDepends.length > 0) svc.depends_on = allDepends;
    return svc;
  },
};

File: cli-go/internal/modules/compose/devcontainer.go
package compose

import (
	"fmt"

	"github.com/joacohbc/my-devcontainer-installer/cli-go/internal/core"
)

const SSHServiceName = "devcontainer-ssh"
const HostDockerSocket = "/var/run/docker.sock"

type DockerSocketMode string

const (
	DockerSocketNone   DockerSocketMode = "none"
	DockerSocketSocket DockerSocketMode = "socket"
	DockerSocketDind   DockerSocketMode = "dind"
)

var DevcontainerService = &ServiceSpec{
	ID:    "devcontainer",
	Label: SSHServiceName + " (main)",
	Always: true,
	Options: []core.ModuleOption{
		{
			ID:    "dockerSocket",
			Label: "Docker access mode",
			Type:  core.ModuleOptionSelect,
			Choices: []core.ModuleOptionChoice{
				{Value: "none", Label: "none — no Docker access (safest)"},
				{Value: "socket", Label: "socket — mount the host /var/run/docker.sock (full host Docker, root-equivalent ⚠️)"},
				{Value: "dind", Label: "dind — isolated rootless Docker-in-Docker engine (sandbox)"},
			},
			Default:        "none",
			RequiresModule: "dod",
		},
	},
	Render: func(ctx RenderContext) *ServiceDef {
		modeStr, _ := ctx.Options["dockerSocket"].(string)
		if modeStr == "" {
			modeStr = "none"
		}
		mode := DockerSocketMode(modeStr)

		dbServices := []string{"mongo", "redis", "postgres"}
		depends := []string{}
		for _, id := range ctx.EnabledServiceIDs {
			for _, db := range dbServices {
				if id == db {
					depends = append(depends, id)
				}
			}
		}

		volumes := []string{
			"../..:/workspace",
			"devcontainer_etc:/etc",
			"devcontainer_root:/root",
			"devcontainer_home:/home",
		}
		if mode == DockerSocketSocket {
			volumes = append(volumes, HostDockerSocket+":"+HostDockerSocket)
		}

		networks := []string{"local-network"}
		if mode == DockerSocketDind {
			networks = append(networks, DindEngineNetwork)
		}

		svc := &ServiceDef{
			Image:         ctx.ImageName,
			Build:         ".",
			ContainerName: SSHServiceName,
			Command:       "sleep infinity",
			Restart:       "unless-stopped",
			Volumes:       volumes,
			Networks:      networks,
		}

		if mode == DockerSocketDind {
			svc.Environment = []string{fmt.Sprintf("DOCKER_HOST=tcp://%s:%d", DindEngineHost, DindEnginePort)}
			depends = append(depends, DindEngineHost)
		}

		if len(depends) > 0 {
			svc.DependsOn = depends
		}

		return svc
	},
}
